CardParts/src/Classes/Card Parts/CardPartProgressBarView.swift

```diff
@@ -0,0 +1,454 @@
+            self.colorBarsStackView.trailingAnchor.constraint(equalTo: self.colorBarsStackViewContainerView.trailingAnchor),
+            self.colorBarsStackView.leadingAnchor.constraint(equalTo: self.colorBarsStackViewContainerView.leadingAnchor),
+            self.colorBarsStackView.topAnchor.constraint(equalTo: self.colorBarsStackViewContainerView.topAnchor),
+            self.colorBarsStackView.bottomAnchor.constraint(equalTo: self.colorBarsStackViewContainerView.bottomAnchor),
+            
+            self.colorBarsStackViewContainerView.heightAnchor.constraint(equalToConstant: self.progressBarOnlyHeight),
+            self.colorBarsStackViewContainerView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor, constant: self.colorsBarOffset),
+            self.colorBarsStackViewContainerView.bottomAnchor.constraint(equalTo: self.valueBarsStackView.topAnchor, constant: -self.colorsBarToValuesBarCushion),
+            
+            self.valueBarsStackView.bottomAnchor.constraint(equalTo: self.view.bottomAnchor),
+            self.valueBarsStackView.heightAnchor.constraint(equalToConstant: self.valuesBarHeight),
+            self.valueBarsStackView.trailingAnchor.constraint(equalTo: self.view.trailingAnchor),
+            self.valueBarsStackView.leadingAnchor.constraint(equalTo: self.view.leadingAnchor)
+        ])
+        
+        super.updateConstraints()
+    }
+    
+    private func layout() {
+        
+        updateConstraints()
+        
+        if let barRadius = barCornerRadius {
+            colorBarsStackView.clipsToBounds = true
+            colorBarsStackViewContainerView.layer.cornerRadius = barRadius
+        }
+        
+        let markerFrame = CGRect(x: self.colorBarsStackView.frame.origin.x, y: 0, width: markerHeight + 3, height: markerHeight)
+        markerView.frame = markerFrame
+        
+        let markerLabelFrame = CGRect(x: 0, y: 0, width: 40, height: self.markerLabelViewHeight)
+        markerLabel.frame = markerLabelFrame
+        markerLabel.text = markerTitle
+        markerLabel.sizeToFit()
+        
+        let viewHeight: CGFloat = valueBarsStackView.frame.height + colorBarsStackView.frame.height + markerView.frame.height + markerLabel.frame.height + markerToColorsBarCushion + colorsBarToValuesBarCushion + labelToMarkerCushion
+        
+        let mainFrame = CGRect(x: self.frame.origin.x, y: self.frame.origin.y, width: self.frame.width, height: viewHeight)
+        self.view.frame = mainFrame
+        self.frame = mainFrame
+        
+        if barValues.count > 1, barColors.count > 0, barValues.count == barColors.count + 1 {
+            animateBasedOnCurrentValue(currentVal)
+        }
+        
+        updateConstraints()
+    }
+    
+    required public init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    
+    fileprivate func addMarker(marker: UIView?) {
```

Comment:
Do we need to deal with removing old markers? What if there is already a marker there?

Reply:
Not required , marker view is optional and  mentioned during it's initialization.

Reply:
No what if they keep calling addMarker... what happens